File: algorithms/39. Combination Sum/solution.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        candidates.sort()  # use sorted in real case
        comb_arr = []
        comb = []
        def dfs(start, target):
            if target == 0:
                comb_arr.append(comb[:])
                return
            for i in range(start, len(candidates)):
                if target < candidates[i]:
                    return
                comb.append(candidates[i])
                dfs(i, target-candidates[i])
                comb.pop()
        dfs(0, target)
        return comb_arr

File: algorithms/39. Combination Sum/solution.cc
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<int> comb;
        vector<vector<int>> result;
        dfs(0, target, candidates, comb, result);
        return result;
    }

    void dfs(int start, int target, vector<int> & nums, vector<int> & comb, vector<vector<int>> & result) {
        if ( target == 0 ) {
            result.push_back(comb);
            return;
        }
        for ( int i = start; i < nums.size() && nums[i] <= target; ++i ) {
            comb.push_back(nums[i]);
            dfs(i, target - nums[i], nums, comb, result);
            comb.pop_back();
        }
    }
};
